File: tree/Traversal/Inorder/iterative/cpp/InorderTraversal.cpp
#include "InorderTraversal.hpp"

#include <vector>

using namespace std;

vector<int> InorderTraversal(const vector<int>& leftChildren, const vector<int>& rightChildren, int root) {
    vector<int> order;
    vector<int> stack;
    int current = root;

    while (current != -1 || !stack.empty()) {
        while (current != -1) {
            stack.push_back(current);
            current = leftChildren[current];
        }

        current = stack.back();
        stack.pop_back();
        order.push_back(current);
        current = rightChildren[current];
    }

    return order;
}

File: tree/Traversal/Inorder/iterative/java/InorderTraversal.java
import java.util.*;

public class InorderTraversal {
    public static List<Integer> InorderTraversal(int[] leftChildren, int[] rightChildren, int root) {
        List<Integer> order = new ArrayList<>();
        Deque<Integer> stack = new ArrayDeque<>();
        int current = root;

        while (current != -1 || !stack.isEmpty()) {
            while (current != -1) {
                stack.push(current);
                current = leftChildren[current];
            }

            current = stack.pop();
            order.add(current);
            current = rightChildren[current];
        }

        return order;
    }
}
